obs-ffmpeg: Enable multi-track audio support for mpegts

pkviet opened the pull request:
### Description
This adds support for multi-track audio for mpegts outputs (srt or rist).
Also modifies UI.
For mpegts compatible protocols, this changes radio buttons into checkboxes for the audio track selection in the Streaming output (only in Advanced output).

![obs64_2023-06-05_21-27-40](https://github.com/obsproject/obs-studio/assets/9283407/473c5eca-d560-4a25-8e11-7c145e3748b0)

### Motivation and Context
YouTube has added support for SRT ingest and allowed broadcast engineers to test it during the 2023 SRT Plugfest.
I was told during the zoom introduction that multi-track support is being worked on by YouTube (public discussion,
there was no NDA).
I had this code around for a while, so it seems it's the right time to add it to obs.
Maybe multi-track will be supported also by RTMP (who knows) and the  code would use the UI parts of this PR.

### How Has This Been Tested?
Tested against several services during 2023 SRT Plugfest.
Works as expected. One can play any track.

### Types of changes
- New feature (non-breaking change which adds functionality) 

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

GeorgesStavracas commented:
We've reviewed this and the question that was raised is if this change prevents unselecting all audio tracks. Recently OBS started enforcing that at least one audio track is selected for recordings, and this behaviour should be reflected here too.

pkviet commented:
> We've reviewed this and the question that was raised is if this change prevents unselecting all audio tracks. Recently OBS started enforcing that at least one audio track is selected for recordings, and this behaviour should be reflected here too.

@GeorgesStavracas i checked current behaviour with recording and there's a warning indeed when no track is selected. I see now what you meant. I've added that too in case some clueless user uncheck all audio tracks.

tytan652 commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -2151,7 +2212,12 @@ bool AdvancedOutput::StartStreaming(obs_service_t *service)
 	bool enableDynBitrate =
 		config_get_bool(main->Config(), "Output", "DynamicBitrate");
-
+	bool is_mpegts = false;
+	obs_service_t *service_obj = main->GetService();
+	const char *url = obs_service_get_connect_info(
+		service_obj, OBS_SERVICE_CONNECT_INFO_SERVER_URL);
+	if (url != NULL)
+		is_mpegts = isMpegts(url);
```
You could just check the protocol that the service return and not the server URL.

pkviet replied:
ah yeah good point. I haven't adjusted yet to the new api

pkviet replied:
done

tytan652 reviewed (commented):
I think it should not be coded as a MPEG TS only feature when, it comes to naming since in the future another container/protocol might do the same thing.

norihiro commented:
The commit 73dd9f82b has a too long line, 80 characters, where it's limit is 72.
> This enables audio multi-track support in UI for mpegts streams (srt, rist ...).

pkviet commented:
> I think it should not be coded as a MPEG TS only feature when it comes to naming, since in the future another container/protocol might do the same thing.

yeah, I had already thought about it. This is code I wrote a long while ago. But given there's a new rtmp spec which is still being worked on, I did plan on updating the code to be protocol neutral. I wanted the PR out sooner than later due to SRT PlugFest.

pkviet commented:
i'm gonna also use dedicated stream tracks instead of the recording tracks since some people might wan tto change encoders for streaming and recording.

pkviet commented:
**Update**
- switched to neutral UI changes to accommodate other protocols with multi-track capability (ex: whip).
- removed the recourse to set_mixers function reserved to raw audio.
- I'll later add support for language tags

krakow10 commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -2060,7 +2114,6 @@ inline void AdvancedOutput::SetupVodTrack(obs_service_t *service)
 			vodTrackEnabled = false;
 	}
-
 	if (vodTrackEnabled && streamTrack != vodTrackIndex)
```
`int streamTrack` in this function is overshadowing `OBSEncoder streamTrack` used in the PR

pkviet replied:
Thanks for catching this. `int streamTrack` is named int streamTrackIndex in other places; so I've renamed it accordingly in the two places where it's used.

krakow10 commented:
I've been using this patch in one form or another since July 2022.  I can confirm that it continues to work on OBS v30.

AgainPsychoX commented:
When can we expect merge? Many streamers want to use this, but they are not technical enough to trust anything other than official build :C

pkviet commented:
> I've been using this patch in one form or another since July 2022.  I can confirm that it continues to work on OBS v30.

Well, I first have to address the comments, which I didn't because ... life.
I hope it can be merged in next point release. We'll see

t3pmd commented:
Would love to see this in the next version please. would help alot

pkviet commented:
Was so long ago, I had forgotten that I had actually addressed the comments. Will request that it moves forward.

pkviet commented:
Rebased to current master.
